src/ssl/support.h

```diff
@@ -273,25 +294,28 @@ bool configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::Po
  */
 void useSquidUntrusted(SSL_CTX *sslContext);
 
-/**
-   \ingroup ServerProtocolSSLAPI
-   * Iterates over the X509 common and alternate names and to see if  matches with given data
-   * using the check_func.
-   \param peer_cert  The X509 cert to check
-   \param check_data The data with which the X509 CNs compared
-   \param check_func The function used to match X509 CNs. The CN data passed as ASN1_STRING data
-   \return   1 if any of the certificate CN matches, 0 if none matches.
- */
-int matchX509CommonNames(X509 *peer_cert, void *check_data, int (*check_func)(void *check_data,  ASN1_STRING *cn_data));
+/// an algorithm for checking/testing/comparing X.509 certificate names
+class GeneralNameMatcher: public Interface
+{
+public:
+    /// whether the given name satisfies algorithm conditions
+    bool match(const Ssl::GeneralName &) const;
 
-/**
-   \ingroup ServerProtocolSSLAPI
-   * Check if the certificate is valid for a server
-   \param cert  The X509 cert to check.
-   \param server The server name.
-   \return   true if the certificate is valid for the server or false otherwise.
- */
-bool checkX509ServerValidity(X509 *cert, const char *server);
+protected:
+    // The methods below implement public match() API for each of the
+    // GeneralName variants. For each public match() method call, exactly one of
+    // these methods is called.
+
+    virtual bool matchDomainName(const Dns::DomainName &) const = 0;
+    virtual bool matchIp(const Ip::Address &) const = 0;
+};
+
+/// Determines whether at least one common or alternate subject names matches.
+/// The first match (if any) terminates the search.
+bool findMatchingSubjectName(X509 &, const GeneralNameMatcher &);
+
+/// whether at least one common or alternate subject name matches the given one
+bool findSubjectName(X509 &, const AnyP::Host &);
```

Comment:
Since we are not really finding the position of the given needle here, we may want to use a hasFoo() naming scheme instead:

```suggestion
bool hasSubjectName(X509 &, const AnyP::Host &);
```

Same for findMatchingSubjectName(), of course.

We should also capitalize these free functions to follow Squid naming guidelines. I will do that once the name is settled.

Reply:
Done in commit a8e992a0.